Update to reduce flakiness of this test set removed flaky annotation

staubina opened the pull request:
TNL-5582
## [TNL-5582](https://openedx.atlassian.net/browse/TNL-5582)
### Description
- Updated the Course updates page to remove unnecessary emptying of the updates list.  This improves the reliability of the page and removes flakiness in the testing 
- Ran the flaky tests 100 times and they passed all 100.
### Sandbox

https://staubina-flaky.sandbox.edx.org
- [x] Build a sandbox for your branch and add a link here
### Testing
- [x] safecommit violation code review process
- [x] Unit, integration, acceptance tests as appropriate
### Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens 
- [x] Code review: @andy-armstrong 
### Post-review
- [x] Rebase and squash commits

staubina commented:
@cahrens Who should be the second reviewer on this PR?

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -18,31 +19,47 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                    // push notification is always disabled for existing updates
-                        var newEle = self.template({updateModel: update, push_notification_enabled: false});
-                        $(updateEle).append(newEle);
+                        debugger;
```
Don't forget to remove debugger

cahrens commented on common/test/acceptance/pages/studio/course_info.py:
```diff
@@ -37,6 +37,8 @@ def click_new_update_button(self):
         """
         Clicks the new-update button.
         """
+        self.wait_for_element_visibility('.course-handouts .edit-button', 'Waiting for edit button to be loaded')
```
Add a comment indicating why you need these waits.

staubina commented:
@andy-armstrong If you have the time would you please review this PR for a Flaky test?

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -1,8 +1,9 @@
+        'edx-ui-toolkit/js/utils/html-utils'],
     function(ValidatingView, CodeMirror, CourseUpdateModel, PromptView, NotificationView,
-        CourseInfoHelper, ModalUtils, DateUtils) {
+        CourseInfoHelper, ModalUtils, DateUtils, HtmlUtils) {
```
Minor Nit: I prefer to order the helper classes earlier in the list, to mirror what we do in Python imports.

staubina replied:
Is `function(ModalUtils, DateUtils, HtmlUtils, ValidatingView, CodeMirror, CourseUpdateModel, PromptView, NotificationView,
        CourseInfoHelper) {` Better?

andy-armstrong replied:
Much better! I'd probably move CodeMirror first as a vendor library, and then I'd move CourseUpdateModel towards the end as it is a model for this code itself.

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -18,31 +19,46 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
+
+            render: function() {
+                // iterate over updates and create views for each using the template
+                var updateEle = this.$el.find('#course-update-list'),
```
Nit: don't use abbreviations, and name all JQuery variables starting with ``$``. I'd call this ``$updateList``.

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -18,31 +19,46 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                        $(updateEle).append(newEle);
+                        // push notification is always disabled for existing updates
+                        HtmlUtils.append(
+                            $(updateEle),
```
Isn't this already a JQuery element? I don't think it needs to be wrapped with ``$(...)``.

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -18,31 +19,46 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                this.$el.find('.new-update-form').hide();
+
+                // Hide Update forms that are not for new updates with the editing class
+                updateEle.children().each(function(index, update) {
```
I found this confusing because ``update`` is used earlier to represent an update model, but here it represents an element. I'd call this one ``updateElement``.

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -18,31 +19,46 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
+
+                // Hide Update forms that are not for new updates with the editing class
+                updateEle.children().each(function(index, update) {
+                    var updateForm = $(update).find('.new-update-form');
```
I'd make a variable for ``$updateElement`` since it is used three times.

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -74,10 +90,14 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                    );
-                        $(ele).find('.date-display').parent().append(this.errorTemplate({message: error[field]}));
+                        HtmlUtils.append(
+                            $(ele).find('#update-date-' + model.cid).parent(),
```
Rename the variable so it doesn't use an abbreviation.

staubina commented:
jenkins run js

andy-armstrong commented on cms/static/js/views/course_info_update.js:
```diff
@@ -257,14 +277,14 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
                     try {
                     // just in case the content causes an error (embedded js errors)
-                        this.$currentPost.find('.update-contents').html(content);
+                        HtmlUtils.setHtml(this.$currentPost.find('.update-contents'), HtmlUtils.HTML(content));
```
Where does ``content`` come from? It would be better to wrap it with ``HTML`` immediately, rather than have consumers remember to do it like this. It makes it clearer that it is HTML at the source. You might also want to name it as ``fooHtml`` (although ``contentHtml`` sounds weird) so that code that consumes it is clear that it contains safely escaped HTML.

staubina commented:
jenkins run python
jenkins run bokchoy

staubina commented:
@andy-armstrong @cahrens 

I have updated my PR with the following changes.

# A Fix for a defect that I had created while testing.
# Updates to the tests to now handle the New Update button being disabled.
# PR and Quality updates for all these changes.

I have also updated the Sandbox.

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -17,32 +23,53 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
 
             initialize: function() {
                 this.template = this.loadTemplate('course_info_update');
-                this.render();
```
So no call to render in initialize because the collection listeners will cause it to be called later?

staubina replied:
Yes.  It was being called twice, which was causing resets when it was not needed.  

@andy-armstrong @cahrens Why is there a need to call render in the init with Backbone?  I tested around this and the page seems to still function as expected.

cahrens replied:
That actually makes me wonder if simply not rendering in initialize is enough to solve the problem (without dealing with disabling the button).

staubina replied:
I feel like this was one of the many things I tried, but I will try it again.

staubina replied:
So this looks to be true, I am just trying to find the right way to ensure the tests will fire as expected and wait for the button.

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -17,32 +23,53 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
+
+                            if (index === self.collection.length - 1) {
+                                // Once the collection is loaded show the button.
+                                self.$el.find('.new-update-button').prop('disabled', false);
```
Why not simplify matters and put this line after line 63 (removing the else clause at line 61)? The try/catch (which was pre-existing) also makes me nervous, because as the code is, there is a chance you would never re-enable this button (if the last update throws an error).

staubina replied:
This needs to happen when the list is completely loaded and this function seems to be running asynchronous to the rest of the code in render.  Therefore when we get to line 63 we have not finished the processing here in this method.   

That being said I could move it to earlier in the same function, but if I pull it out I believe I will have the same issue as before, New update button enabled before the collection was loaded.  Lastly, if I move the code on 63 outside the else it will enable the button before the collection is loaded.

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -70,31 +97,35 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                var ele = this.$el.find('#course-update-list li[name=\"' + model.cid + '\"]');
-                $(ele).find('.message-error').remove();
+                var validationElement = this.$el.find('#course-update-list li[name="' + model.cid + '"]');
+                $(validationElement).find('.message-error').remove();
```
And there are a couple more locations below.

cahrens commented on cms/static/js/views/course_info_update.js:
```diff
@@ -70,31 +97,35 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                    $(ele).find('.message-error').remove();
-                    $(ele).find('.save-button').removeClass('is-disabled');
+                    $(validationElement).find('.message-error').remove();
+                    $(validationElement).find('.save-button').removeClass('is-disabled');
```
I think we will have to revisit this code in the future to make the save button _actually_ disabled (instead of just CSS class). But out of scope for this change.

staubina replied:
I am unable to click the save button when I make an error, but we can absolutely revisit this issue.

cahrens replied:
It may still be possible to tab to it (that's what we've seen in the LMS). But again, at this point it doesn't really matter.

cahrens commented:
@staubina done reviewing. There is some more simplification that can be done, but overall looking good.

staubina commented:
jenkins run python

staubina commented on cms/static/js/views/course_info_update.js:
```diff
@@ -17,32 +23,53 @@ define(['js/views/validation', 'codemirror', 'js/models/course_update',
-                this.render();
-            // when the client refetches the updates as a whole, re-render them
+                // Disable the new update button while we load the collection.
+                this.$el.find('.new-update-button').prop('disabled', true);
```
I attempted this initially and was unable to remove the disabled state.  I will try again now that my changes have stabilized.

staubina replied:
Please stop by when you have a moment.  This is not working for me as I would expect it to.

cahrens replied:
How are you specifying disabled in the HTML? See http://www.w3schools.com/tags/att_disabled.asp for the correct way of specifying it.

staubina commented:
I am running the tests again through the test runner.  I will post here when it is successful.

staubina commented:
@cahrens @andy-armstrong Sandbox updated and requested changes made.

staubina commented:
jenkins run bokchoy

cahrens commented on common/test/acceptance/pages/studio/course_info.py:
```diff
@@ -41,11 +41,9 @@ def is_update_button_enabled():
-            if update_button_prop is not None:
-                return len(update_button_prop) == 0
-            else:
+            if self.q(css='.new-update-button').attrs('disabled')[0] is None:
```
Nit: can be simplified as

````
return self.q(css='.new-update-button').attrs('disabled')[0] is None

staubina replied:
Will do, I was just running to tests with the Jenkins tester to verify it is working and not flaky.

staubina commented:
@cahrens  I made the test change. Would you like me to update the Sandbox again?

staubina commented:
@cahrens Should I squash commits?

cahrens commented on cms/templates/course_info.html:
```diff
@@ -45,7 +45,7 @@ <h1 class="page-header">
         <ul>
           <li class="nav-item">
-            <a href="#" class=" button new-button new-update-button"><span class="icon fa fa-plus" aria-hidden="true"></span> ${_('New Update')}</a>
+            <a href="#" class=" button new-button new-update-button" disabled><span class="icon fa fa-plus" aria-hidden="true"></span> ${_('New Update')}</a>
```
To clarify-- waiting until a-links are converted to buttons throughout Studio (which is a pervasive issue and will require CSS changes too).

cahrens commented:
👍 Yes, please squash.

staubina commented:
jenkins run python
